Update `push` and `pull` docs (#2685)
jenssegers/laravel-mongodb: docs/query-builder.md

**Push**

Add one or multiple values to the items array.

```php
// Push the value to the matched documents
DB::collection('users')
  ->where('name', 'John')
    // Push a single value to the items array
  ->push('items', 'boots'); // items: ['boots']

DB::collection('users')
  ->where('name', 'John')
    // Push multiple values to the items array
  ->push('items', ['hat', 'jeans']); // items: ['boots', 'hat', 'jeans']

// Or

// Push the value to an exact document
$user->push('items', 'boots');
$user->push('items', ['hat', 'jeans']);
```

> To add an array to the messages array.

Reviewer: We should add a note that values arrays must be nested into a list array, otherwise they are interpreted as a list of scalar values.

```php
DB::collection('users')
  ->where('name', 'John')
    // Push an array as a value to the items array
  ->push('messages', [
      [
          'from'    => 'Jane Doe',
          'message' => 'Hi John',
      ],

>   ]); // messages: [ ['from' => 'Jane Doe', 'message' => 'Hi John'] ]

Reviewer: This should be presented as BSON documents: (I'm not sure about the comment style, it might be better to use a separate the code block with `ts` syntax.

```suggestion
  ]);
  // Result:
  // messages: [
  //     { from: "Jane Doe", message: "Hi John" }
  // ]
```

Same for other examples.

**Push**

Add one or multiple values to the `items` array.

```php
// Push the value to the matched documents
DB::collection('users')
  ->where('name', 'John')
  // Push a single value to the items array
  ->push('items', 'boots');
// Result:
// items: ['boots']

DB::collection('users')
  ->where('name', 'John')
  // Push multiple values to the items array
  ->push('items', ['hat', 'jeans']);
// Result:
// items: ['boots', 'hat', 'jeans']

// Or

// Push the values directly to a model object
$user->push('items', 'boots');
$user->push('items', ['hat', 'jeans']);
```

To add embedded document or array values to the `messages` array, those values must be specified within a list array.

```php
DB::collection('users')
  ->where('name', 'John')
    // Push an embedded document as a value to the messages array
  ->push('messages', [

>       [ 'from'    => 'Jane Doe', 'message' => 'Hi John' ]

Reviewer 2: ```suggestion
      [ 'from' => 'Jane Doe', 'message' => 'Hi John' ]
```

  ]);
// Result:
// messages: [
//      { from: "Jane Doe", message: "Hi John" }
//  ]

// Or

$user->push('messages', [

>     [ 'from'    => 'Jane Doe', 'message' => 'Hi John' ]

Reviewer 2: ```suggestion
    [ 'from' => 'Jane Doe', 'message' => 'Hi John' ]
```

Not sure where this extra indentation came from.

]);
```

If you **DON'T** want duplicate values, set the third parameter to `true`:

```php
DB::collection('users')
  ->where('name', 'John')
  ->push('items', 'boots');
// Result:
// items: ['boots']

DB::collection('users')
  ->where('name', 'John')
  ->push('items', ['hat', 'boots', 'jeans'], true);
// Result:
// items: ['boots', 'hat', 'jeans']

// Or

$user->push('messages', [

>     [ 'from'    => 'Jane Doe', 'message' => 'Hi John' ]

Reviewer 2: ```suggestion
    [ 'from' => 'Jane Doe', 'message' => 'Hi John' ]
```

]);
// Result:
// messages: [
//      { from: "Jane Doe", message: "Hi John" }
//  ]

$user->push('messages', [

>     [ 'from'    => 'Jess Doe', 'message' => 'Hi' ],
>     [ 'from'    => 'Jane Doe', 'message' => 'Hi John' ],

Reviewer 2: ```suggestion
    [ 'from' => 'Jess Doe', 'message' => 'Hi' ],
    [ 'from' => 'Jane Doe', 'message' => 'Hi John' ],
```

], true);
// Result:
// messages: [
//      { from: "Jane Doe", message: "Hi John" }
//      { from: "Jess Doe", message: "Hi" }
//  ]
```

**Pull**

Remove one or multiple values from the `items` array.

```php
// items: ['boots', 'hat', 'jeans']

DB::collection('users')
  ->where('name', 'John')
  ->pull('items', 'boots'); // Pull a single value
// Result:
// items: ['hat', 'jeans']

// Or pull multiple values

$user->pull('items', ['boots', 'jeans']);
// Result:
// items: ['hat']
```

> Embedded document and arrays values can also be pulled from the `messages` array.

Reviewer 2: ```suggestion
Embedded document and arrays values can also be removed from the `messages` array.
```

I think "removed" is clearer to the reader here, since you use the same term in the text above the preceding example.

```php
// Latest state:
// messages: [
//      { from: "Jane Doe", message: "Hi John" }
//      { from: "Jess Doe", message: "Hi" }
//  ]

DB::collection('users')
  ->where('name', 'John')

>     // It will search for the exact array and pulls it

Reviewer 2: ```suggestion
    // Pull an embedded document from the array
```

I don't think we need to say "exact" here, as `$pull` performs a match. Quoting [`$pull`: Behavior](https://www.mongodb.com/docs/manual/reference/operator/update/pull/#behavior): 

> If the specified <value> to remove is a document, [$pull](https://www.mongodb.com/docs/manual/reference/operator/update/pull/#mongodb-update-up.-pull) removes only the elements in the array that have the exact same fields and values. The ordering of the fields can differ.

MongoDB still has logic elsewhere where BSON field order factors into comparisons (as I mentioned in https://github.com/mongodb/laravel-mongodb/pull/2685/files#r1406281116), but it's thankfully not a concern with `$pull`. I believe [`$pullAll`](https://www.mongodb.com/docs/manual/reference/operator/update/pullAll/#behavior) is what actually does an exact match with respect to field order. I pray there isn't an extra boolean argument on the `pull()` method that determines which operator to use, so hopefully this won't come up :laughing: 

----

Example from the shell:

```
> db.foo.find()
[
  {
    _id: ObjectId("656758ae5b133c573d501330"),
    x: 1,
    y: [ { a: 1, b: 2, c: 3 } ]
  }

> db.foo.updateOne({ x: 1 }, { $pullAll: { y: [ { a: 1 }]}})
{
  acknowledged: true,
  insertedId: null,
  matchedCount: 1,
  modifiedCount: 0,
  upsertedCount: 0
}

> db.foo.updateOne({ x: 1 }, { $pull: { y: {a: 1}}})
{
  acknowledged: true,
  insertedId: null,
  matchedCount: 1,
  modifiedCount: 1,
  upsertedCount: 0
}

> db.foo.find()
[ { _id: ObjectId("656758ae5b133c573d501330"), x: 1, y: [] } ]
```

  ->pull('messages', [
      [ 'from' => 'Jane Doe', 'message' => 'Hi John' ]
  ]);
// Result:
// messages: [
//      { from: "Jess Doe", message: "Hi" }
//  ]

> // Or pull multiple array

Reviewer 2: ```suggestion
// Or pull multiple embedded documents
```

$user->pull('messages', [
    [ 'from' => 'Jane Doe', 'message' => 'Hi John' ],

>     [ 'from' => 'Jess Doe', 'message' => 'Hi' ]

Reviewer 2: Again, the field order shouldn't matter here if the library is always using `$pull`.

Author: I see. Thanks.

**Push**

> Add one or multiple values to the items array.

Reviewer 2: ```suggestion
Add one or multiple values to the `items` array.
```

Quoting field names should make this more readable.

```php
// Push the value to the matched documents
DB::collection('users')
  ->where('name', 'John')
  // Push a single value to the items array
  ->push('items', 'boots');
// Result:
// items: ['boots']

DB::collection('users')
  ->where('name', 'John')
  // Push multiple values to the items array
  ->push('items', ['hat', 'jeans']);
// Result:
// items: ['boots', 'hat', 'jeans']

// Or

// Push the value to an exact document
$user->push('items', 'boots');

> $user->push('items', ['hat', 'jeans']);

Reviewer 2: > Push the value to an exact document

The "exact document" language isn't very clear to me. I assume the difference here is that it's referring to calling `push()` on a model object instead of executing an update through the collection.

Would "Push the values directly to a model object" read better?

Author: Sorry, I'm not that good in English to write docs😅

```

> To add an array as a value to the messages array. value arrays must be nested into a list array, otherwise they are
> interpreted as a list of scalar values.

Reviewer 2: ```suggestion
To add embedded document or array values to the `messages` array, those values must be specified within a list array.
```

> otherwise they are interpreted as a list of scalar values.

I think this raises more questions, as it's unclear what would happen if the second argument to `push()` was an associative array. I'd hope there'd be an exception thrown if the argument is not a real list, but I wouldn't be surprised if it was undefined behavior (maybe keys are ignored).

Offhand, does the method internally wrap its arguments in an [`$each`](https://www.mongodb.com/docs/manual/reference/operator/update/each/#mongodb-update-up.-each) operator?

Author: https://github.com/mongodb/laravel-mongodb/blob/973b6b90f7916916bf252ebd707fe4f4b5eaaeb1/src/Eloquent/Model.php#L470

Yes, It does.

```php
DB::collection('users')
  ->where('name', 'John')

>     // Push an array as a value to the items array

Reviewer 2: ```suggestion
    // Push an embedded document as a value to the items array
```

Using the correct term (from MongoDB's POV) here will help avoid any confusion.

  ->push('messages', [
      [
          'from'    => 'Jane Doe',
          'message' => 'Hi John',
      ],
  ]);
// Result:
// messages: [
//      { from: "Jane Doe", message: "Hi John" }
//  ]

// Or

$user->push('messages', [
    [
        'from'    => 'Jane Doe',
        'message' => 'Hi John',
    ],
]);
```

> If you **DON'T** want duplicate values, set the third parameter to `true`:

Reviewer 2: I haven't checked the code, but this suggests that the boolean causes [`$addToSet`](https://www.mongodb.com/docs/manual/reference/operator/update/addToSet/) to be used internally instead of [`$push`](https://www.mongodb.com/docs/manual/reference/operator/update/push/). Is that correct?

Having a boolean argument control that rubs me the wrong way (I'll add it to the growing list), but I assume there's an explanation for this behavior. Is the `push()` method here actually derived from something else in Eloquent, or was it originally created in this library specifically for the MQL operators?

Author: > I haven't checked the code, but this suggests that the boolean causes [`$addToSet`](https://www.mongodb.com/docs/manual/reference/operator/update/addToSet/) to be used internally instead of [`$push`](https://www.mongodb.com/docs/manual/reference/operator/update/push/). Is that correct?

Yes. It's the logged queries of pushing normally and pushing unique values.

```plain
array:2 [
  0 => array:3 [
    "query" => "users.updateMany({"_id":"65672f58551357588a07c993"},{"$push":{"items":{"$each":["boots"]}}},{"multiple":true})"
    "bindings" => []
    "time" => 1.82
  ]
  1 => array:3 [
    "query" => "users.updateMany({"_id":"65672f58551357588a07c993"},{"$addToSet":{"items":{"$each":["hat","jeans"]}}},{"multiple":true})"
    "bindings" => []
    "time" => 0.6
  ]
]
```
 

> Having a boolean argument control that rubs me the wrong way (I'll add it to the growing list), but I assume there's an explanation for this behavior. Is the `push()` method here actually derived from something else in Eloquent, or was it originally created in this library specifically for the MQL operators?

 The `push` method is originally from Eloquent and it's used for saving a model instance and its relationships to the database. this library overrides the `push` method and accepts three args. If no arg is passed, the original `push` method will be called.

Reviewer 2: Is Eloquent's `push` method documented anywhere beyond [Recursively Saving Models & Relationships](https://laravel.com/docs/10.x/eloquent-relationships#the-push-method)? That doesn't appear to take _any_ arguments and the method seems entirely different in purpose (this doesn't have to do with relationships).

@GromNaN: A boolean flag to determine whether `$push` or `$addToSet` is used seems like a poor design as it makes the code less self-documenting. I think this is compounded by the fact that the method on the model class [uses `func_get_args()`](https://github.com/mongodb/laravel-mongodb/blob/973b6b90f7916916bf252ebd707fe4f4b5eaaeb1/src/Eloquent/Model.php#L459C23-L459C36) and doesn't define a named parameter that users could specify if they so desired.

It may be worth making a ticket to revisit this down the line. Perhaps introduce a new `addToSet()` method and then deprecate the original variation. This might also be a moot point when/if the query builder is integrated.

Reviewer: 💯 agree that the double behavior of this method is a bad design.

If deprecate the custom behavior of `Model::push` with parameters, we have to do deprecate the `pull` method too.

Let's improve the documentation for now.

```php
DB::collection('users')
  ->where('name', 'John')
  ->push('items', 'boots');
// Result:
// items: ['boots']

DB::collection('users')
  ->where('name', 'John')
  ->push('items', ['hat', 'boots', 'jeans'], true);
// Result:
// items: ['boots', 'hat', 'jeans']

// Or

$user->push('messages', [
    [
        'from'    => 'Jane Doe',
        'message' => 'Hi John',
    ],
]);
// Result:
// messages: [
//      { from: "Jane Doe", message: "Hi John" }
//  ]

$user->push('messages', [
    [
        'from'    => 'Jess Doe',
        'message' => 'Hi',
    ],
    [
        'from'    => 'Jane Doe',
        'message' => 'Hi John',
    ],

> ], true);

Reviewer 2: This example may open a can of worms about how MongoDB compares documents. For instance, if you happened to swap the order of `from` and `messages` here, the document would likely be appended and you'd end up with three elements in the `messages` array. [Comparison and Sort Order: Objects](https://www.mongodb.com/docs/manual/reference/bson-type-comparison-order/#objects) goes into more detail about that.

If it's not too useful to readers, I'd propose removing it. An earlier example already demonstrates that we can call `push()` on a model object directly, so I wouldn't push for repeating that multiple times (pun intended).

Author: > If it's not too useful to readers, I'd propose removing it.

It's the reason that I've opened this PR. Users get confused when they want to use these methods (like this issue #2387). To reduce the complexity here, I suggest using arrays with only one key and value.

```suggestion
$user->push('messages', [
    [ 'message' => 'Hi' ],
    [ 'message' => 'Hi John' ],
], true);
```

Reviewer 2: Ah, just saw your comment in https://github.com/mongodb/laravel-mongodb/issues/2387#issuecomment-1819110263. In that case, I think it's fine to leave the embedded documents as-is with two fields. Putting myself in the reader's shoes for a moment, I think seeing an embedded document with a single string field would make me wonder why the author didn't just store `messages` as a list of strings.

// Result:
// messages: [
//      { from: "Jane Doe", message: "Hi John" }
//      { from: "Jess Doe", message: "Hi" }
//  ]
```

**Pull**

> Remove one or multiple values form items array.

Reviewer 2: ```suggestion
Remove one or multiple values from the `items` array.
```

```php
// items: ['boots', 'hat', 'jeans']

DB::collection('users')
  ->where('name', 'John')
  ->pull('items', 'boots'); // Pull a single value
// Result:
// items: ['hat', 'jeans']

// Or pull multiple values

$user->pull('items', ['boots', 'jeans']);
// Result:
// items: ['hat']
```

> Also, you can pull an array value from the messages array.

Reviewer 2: ```suggestion
Embedded document and arrays values can also be pulled from the `messages` array.
```

Using the specific nomenclature here.

Avoiding "you" in documentation when possible is also a plus. The PHP docs are pretty good about this (see: [Style Guidelines: Personalization](http://doc.php.net/tutorial/style.php) for their reasoning).

Author: Thank you, Jeremy. I will keep this in mind.

```php
// Latest state:
// messages: [
//      { from: "Jane Doe", message: "Hi John" }
//      { from: "Jess Doe", message: "Hi" }
//  ]

DB::collection('users')
  ->where('name', 'John')
    // It will search for the exact array and pulls it
  ->pull('messages', [
      [
          'from' => 'Jane Doe',
          'message' => 'Hi John',

>       ]

Reviewer 2: Consider writing all of these embedded documents as one-liners through this PR to make the examples more concise. This would be particularly helpful for the example where a third boolean argument is provided to `push()` since it can easily get lost at the end of all of the whitespace in the method call.

          'from'    => 'Jane Doe',
          'message' => 'Hi John',
      ],
  ]);
// Result:
// messages: [

> //      { 'from' => 'Jane Doe', 'message' => 'Hi John' }

Reviewer: This is half way to a JS format.
```suggestion
//      { from: "Jane Doe", message: "Hi John" }
```
